### DISCUSSION TITLE
End2End training Idea - one branch - decay topk from 10 to 1

### DISCUSSION THREAD

**Discussion by pfeatherstone** [ultralytics/ultralytics#23299] [Ideas]
My understanding of end2end training is you have:
* one2many : cv2 and cv3 layers in Detect
* one2one: one2one_cv2 and one2one_cv3 layers

You have two v8Detection losses, one for one2many and another for one2one. The former uses TAL with topk == 10 and the latter uses TAL with topk == 1.

You minimize both losses starting with initial weights of 0.8 and 0.2 respectively which are then decayed and inversely decayed over time. So towards the end of training, the one2one branch gets most of the weighting.

Here is the idea, instead of having two branches, simply have one and decay topk from 10 to 1?
So you're decaying the one2many behaviour to one2one. That way you only have one branch. You save memory. Maybe that works.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @pfeatherstone, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Thanks for sharing this End2End training idea 💡—it’s a thoughtful direction, and it’d be super helpful to understand your target setup and what you’d like to optimize (VRAM, speed, accuracy tradeoffs, etc.) 🔍  
Note: this is an automated response, and an Ultralytics engineer will also assist soon 🙂

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by glenn-jocher** [reply]
You’re reading the setup correctly (two heads: `cv2/cv3` and `one2one_cv2/one2one_cv3`), but note our “decay” implementation today is on the **loss weights** (see `E2ELoss.update()`/`decay()`), while the one-to-one assigner is not strictly “topk=1” in all cases (e.g. `E2ELoss` builds one2one with `tal_topk=7, tal_topk2=1`). See the current implementations in the [loss reference (`ultralytics/utils/loss.py`)](https://docs.ultralytics.com/reference/utils/loss/) and the [Detect head reference (`ultralytics/nn/modules/head.py`)](https://docs.ultralytics.com/reference/nn/modules/head/).  

Decaying `tal_topk` from 10→1 on a single branch is an interesting experiment, but it won’t be equivalent to two-branch training because the assignment is a discrete switch and the separate heads let the two objectives specialize; also the duplicated head params are usually a small fraction of total VRAM vs backbone/neck activations—if you do try it, a small PR with COCO metrics + VRAM numbers would be the best way to evaluate the tradeoff.

**Reply by pfeatherstone** [reply]
It seems a bit overkill to me training two heads. I can see why. The one2many branch gives more "signal" to the gradients and helps with convergence early on. But the two branches are completely independent. They are not tied in any way. It's not like one2one_cv2 is a pruned version of cv2 and likewise for one2one_cv3 and cv3. They are trained independently. So you're basically training two networks where backbone and fpn are shared. I wonder if the same result could be achieved with training one branch and topk==10. Then fine tune at the end with topk==1. Or maybe somehow prune cv2/cv3 first, then fine tune. Or, like i suggested, decay topk.

**Reply by pfeatherstone** [reply]
I can see that even with end2end training, a poorly trained model might still output multiple overlapping boxes and might still need NMS. Removing NMS completely only works if you have a really well trained model. There's nothing inherent about end2end training that forces 1 detection per object if that makes sense. We just hope it does because of the topk==1 training. I originally thought end2end training did something like bipartite matching in DETR and did direct set prediction. But it's none of that correct? You just threshold the detections by confidence and hope that all your predictions are "unique".

**Reply by glenn-jocher** [reply]
Good points, but one correction: in our end2end `Detect.forward()` the one-to-one head is run on `x_detach = [xi.detach() for xi in x]`, so the one2one loss updates only the one2one head params (no backbone/neck grads), which is why it’s not really “two full networks” in training ([`Detect.forward()` in `ultralytics/nn/modules/head.py`](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/nn/modules/head.py#L141-L155)). Your “train topk=10 then finetune topk=1 / decay topk” ideas are reasonable experiments, but they won’t be equivalent unless you also replicate the one2one assignment behavior we use (note `E2ELoss` builds one2one with `tal_topk=7, tal_topk2=1`, not a pure topk=1) ([`E2ELoss` init in `ultralytics/utils/loss.py`](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/loss.py#L1129-L1161)).  

On “NMS-free”: it’s not DETR-style Hungarian matching, but inference is also not “threshold and hope” either—we explicitly take the top scoring predictions up to `max_det` (see the top-k selection in [`Detect.get_topk_index()`](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/nn/modules/head.py#L219-L237)); if a model is undertrained you can still see duplicates, and in that case external NMS is a valid fallback, but the training objective is designed to make the one2one head produce a single dominant prediction per object.

**Reply by pfeatherstone** [reply]
Sorry, I didn't spot the .detach() for one2one branch. 
Thanks for your comments. Really interesting. I'll have a closer look at TAL when topk2 is defined.

**Reply by glenn-jocher** [reply]
No worries — `tal_topk`/`tal_topk2` ultimately flow into the Task-Aligned Assigner, so the most direct place to see exactly how the second-stage selection works is the [TaskAlignedAssigner implementation (`ultralytics/utils/tal.py`)](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/tal.py); if you try a scheduled `tal_topk` decay (or staged finetune) and can share COCO metrics + VRAM deltas, that’d be the cleanest way for us to evaluate the tradeoff.

**Reply by pfeatherstone** [reply]
it says in the docs you completely removed DFL loss. It looks like however you replaced it with an **L1 loss** between the predicted and ground truth ltrb distances to anchor points. What was the reasoning behind this? Why not just use CIOU loss? You're using both CIOU and L1

**Reply by glenn-jocher** [reply]
Yep — with Ultralytics YOLO26 we drop DFL’s distribution-style box regression (and its extra `reg_max` channels) and instead supervise the direct box outputs; the added L1 term on the raw `ltrb` distances is mainly there to provide a dense, stable gradient even when boxes don’t overlap (IoU/CIoU can be near-zero early on), while CIoU remains the “shape + overlap” objective once predictions are in the right neighborhood, and the combo empirically converges more reliably while keeping the head/export path simpler for end-to-end, NMS-free inference (see [why YOLO26 removes NMS and how that changes deployment](https://www.ultralytics.com/blog/why-ultralytics-yolo26-removes-nms-and-how-that-changes-deployment)).

**Reply by pfeatherstone** [reply]
With the removal of DFL, did you consider adding an activation to the ltrb distances? They could be negative during training which doesn't really make sense. Maybe adding softplus() which guarantees they are positive ? I guess you trained super duper models so demonstrably you didn't need to...

**Reply by glenn-jocher** [reply]
Negative `l/t/r/b` isn’t inherently invalid here—those values are effectively *signed* offsets from the anchor point, so a negative “left” (for example) just means the anchor lies to the left of the predicted box edge (i.e., the anchor isn’t required to be inside the box), and the decode path can still produce a valid `xyxy` box (see the decode flow in the [Detect head reference](https://docs.ultralytics.com/reference/nn/modules/head/)).  

You *can* force positivity with something like `softplus()`, but it tends to bias/slow box regression (especially early training) and can make it harder to represent tight boxes when the model is still “searching,” which is why we generally leave it unconstrained and let the regression + IoU terms shape it. If you want to experiment, apply `F.softplus()` only when you’re in the “no-DFL / `reg_max=1`” regime (otherwise you’d break the distribution-style logits).
